Fix using limit=0 and details=on to get only the count of elements #1492

Anjali-NEC opened the pull request:
Fix issue #1492

fgalan commented on CHANGES_NEXT_RELEASE:
```diff
@@ -1,3 +1,4 @@
+- Fix: Using limit=0 and details=on to get only the count of elements (#1492)
```
```suggestion
- Fix: Using limit=0 and options=count to get only the count of elements (#1492)
```

Anjali-NEC replied:
Fixed in latest commit

fgalan commented:
This PR should include .test files to cover the cases described in https://github.com/telefonicaid/fiware-orion/issues/1492#issuecomment-384898462

In addition, some existing .test are failing. They has to be reviewed and fixed.

Anjali-NEC commented:
> This PR should include .test files to cover the cases described in [#1492 (comment)](https://github.com/telefonicaid/fiware-orion/issues/1492#issuecomment-384898462)
> 
> In addition, some existing .test are failing. They has to be reviewed and fixed.

Added new .test file and fixed existing .test are which are failing.

fgalan commented:
Reviwing this PR I have realized that in line L161

```
        OrionError error(SccBadRequest, std::string("Bad pagination limit: /") + value + "/ [must be a decimal number]");
```

should be changed to a more precise message:

```
        OrionError error(SccBadRequest, std::string("Bad pagination limit: /") + value + "/ [must be a positive integer number]");
```

Could you do that change in the PR, please?

fgalan commented on src/lib/rest/rest.cpp:
```diff
@@ -177,9 +177,7 @@ static int uriArgumentGet(void* cbDataP, MHD_ValueKind kind, const char* ckey, c
     else if (limit == 0)
     {
-      OrionError error(SccBadRequest, std::string("Bad pagination limit: /") + value + "/ [a value of ZERO is unacceptable]");
-      ciP->httpStatusCode = error.code;
-      ciP->answer         = error.smartRender(ciP->apiVersion);
+      ciP->httpStatusCode = SccOk;
       return MHD_YES;
     }
```
It would be cleaner to remove the whole `else if (limit == 0)` block.

Anjali-NEC replied:
Fix in https://github.com/telefonicaid/fiware-orion/pull/3994/commits/8879a3975cdd73efa2a2d9289e528d99c68ff168

fgalan commented on test/functionalTest/cases/0000_bad_requests/pagination_error_in_uri_params.test:
```diff
@@ -93,16 +93,16 @@ Date: REGEX(.*)
+    "errorCode": {
         "code": "400",
-        "details": "Bad pagination limit: /000000/ [a value of ZERO is unacceptable]",
+        "details": "JSON Parse Error",
```
"JSON Parse Error" doesn't seem to be a right message in this case, as actually the request in step 3 doesn't have any payload...

In fact, a OK response seems to be better.

Anjali-NEC replied:
> "JSON Parse Error" doesn't seem to be a right message in this case, as actually the request in step 3 doesn't have any payload...
> 
> In fact, a OK response seems to be better.

In step 3 everything is OK but it shows "JSON Parse Error". I didn't find reason for this error. Could you please suggest what modification should be done in this test case?

fgalan replied:
I understand that the response in this case should be the same you get when do a /v1/queryContext operation without any matching result. You can test yourself using this curl:

```
curl -vvv localhost:1026/v1/queryContext -H 'content-type: application/json' -d @- <<EOF
{
    "entities": [
        {
            "type": "foo",
            "isPattern": "false",
            "id": "bar"
        }
    ]
}
EOF
```

You should get something like this:

```
< HTTP/1.1 200 OK
< Connection: Keep-Alive
< Content-Length: 70
< Content-Type: application/json
< Fiware-Correlator: aaee0e26-4621-11ec-9ae3-000c29583ca5
< Date: Mon, 15 Nov 2021 14:38:15 GMT
< 
* Connection #0 to host localhost left intact
{"errorCode":{"code":"404","reasonPhrase":"No context element found"}}
```

The step in the step should resemble that. The reason behind the "JSON Parser Error" in this case should be analyzed.

fgalan commented on test/functionalTest/cases/1492_using_limit=0_and_options=count_to_get_only_the_count_of_elements/using_limit=0_and_options=count_to_get_only_the_count_of_elements.test:
```diff
@@ -0,0 +1,79 @@
+--SHELL--
+
+#
+# 01. GET /v2/entities with 'limit' == 0 and options=count
+# 02. GET /v2/entities with 'limit' == 0
```
I would suggest to populate some entities to see its effect in `fiware-total-count`. Thus, using a sequence of steps like this:

```suggestion
# 01. GET /v2/entities with 'limit' == 0 and options=count and see no entities and fiware-total-count: 0
# 02. POST /v2/entities to create E1 entity
# 03. POST /v2/entities to create E1 entity
# 04. GET /v2/entities with 'limit' == 0 and options=count and see no entities and fiware-total-count: 2
# 05. GET /v2/entities with 'limit' == 0 and see no entities
```

Anjali-NEC replied:
Added in https://github.com/telefonicaid/fiware-orion/pull/3994/commits/8879a3975cdd73efa2a2d9289e528d99c68ff168

fgalan commented on test/functionalTest/cases/1492_using_limit=0_and_options=count_to_get_only_the_count_of_elements/using_limit=0_and_options=count_to_get_only_the_count_of_elements.test:
```diff
@@ -0,0 +1,79 @@
+# Copyright 2021 Telefonica Investigacion y Desarrollo, S.A.U
```
New name suggestion (to avoid using = character in the filename and make it much simpler):

test/functionalTest/cases/1492_limit_zero/limit_zero_entities.test

fgalan replied:
In addition, I'd suggest to do a couple more tests, to cover subscriptions and registrations pagingation:

test/functionalTest/cases/1492_limit_zero/limit_zero_subs.test
test/functionalTest/cases/1492_limit_zero/limit_zero_regs.test

Same steps that limit_zero_entities.test (see the steps suggested here: https://github.com/telefonicaid/fiware-orion/pull/3994#discussion_r748310446) but using subs and regs instead of subs.

Anjali-NEC replied:
Added in https://github.com/telefonicaid/fiware-orion/pull/3994/commits/8879a3975cdd73efa2a2d9289e528d99c68ff168

Anjali-NEC commented:
> Reviwing this PR I have realized that in line L161
> 
> ```
>         OrionError error(SccBadRequest, std::string("Bad pagination limit: /") + value + "/ [must be a decimal number]");
> ```
> 
> should be changed to a more precise message:
> 
> ```
>         OrionError error(SccBadRequest, std::string("Bad pagination limit: /") + value + "/ [must be a positive integer number]");
> ```
> 
> Could you do that change in the PR, please?

Fix in https://github.com/telefonicaid/fiware-orion/pull/3994/commits/8879a3975cdd73efa2a2d9289e528d99c68ff168

fgalan commented on test/functionalTest/cases/1492_limit_zero/limit_zero_regs.test:
```diff
@@ -0,0 +1,267 @@
+--SHELL--
+
+#
+# 01. GET /v2/registrations with 'limit' == 0 and options=count and see no entities and fiware-total-count: 0
+# 02. POST /v2/registrations to create E1 entity
+# 03. POST /v2/registrations to create E2 entity
+# 04. GET /v2/registrations with 'limit' == 0 and options=count and see no entities and fiware-total-count: 2
+# 05. GET /v2/registrations with 'limit' == 0 and see no entities
```
```suggestion
# 01. GET /v2/registrations with 'limit' == 0 and options=count and see no registrations and fiware-total-count: 0
# 02. POST /v2/registrations to create registration
# 03. POST /v2/registrations to create another registration
# 04. GET /v2/registrations with 'limit' == 0 and options=count and see no registrations and fiware-total-count: 2
# 05. GET /v2/registrations with 'limit' == 0 and see no registration
```

Same changes in echo statements and expectations in the .test file.

Anjali-NEC replied:
Fixed in https://github.com/telefonicaid/fiware-orion/pull/3994/commits/cf68c96ee05fa0fd7b53156f8e2b86761a386d0f

fgalan commented on test/functionalTest/cases/1492_limit_zero/limit_zero_subs.test:
```diff
@@ -0,0 +1,299 @@
+--SHELL--
+
+#
+# 01. GET /v2/subscriptions with 'limit' == 0 and options=count and see no entities and fiware-total-count: 0
+# 02. POST /v2/subscriptions to create E1 entity
+# 03. POST /v2/subscriptions to create E2 entity
+# 04. GET /v2/subscriptions with 'limit' == 0 and options=count and see no entities and fiware-total-count: 2
+# 05. GET /v2/subscriptions 'limit' == 0 and see no entities
```
```suggestion
# 01. GET /v2/subscriptions with 'limit' == 0 and options=count and see no subscriptions and fiware-total-count: 0
# 02. POST /v2/subscriptions to create subscription
# 03. POST /v2/subscriptions to create another subscription
# 04. GET /v2/subscriptions with 'limit' == 0 and options=count and see no subscriptions and fiware-total-count: 2
# 05. GET /v2/subscriptions 'limit' == 0 and see no subscriptions 
```

Same changes in echo statements and expectations in the .test file.

Anjali-NEC replied:
Fixed in https://github.com/telefonicaid/fiware-orion/pull/3994/commits/cf68c96ee05fa0fd7b53156f8e2b86761a386d0f

fgalan commented on test/functionalTest/cases/1492_limit_zero/limit_zero_entities.test:
```diff
@@ -0,0 +1,160 @@
+
+
+
+04. GET /v2/entities with 'limit' == 0 and options=count and see no entities and fiware-total-count: 2
```
This step (and also step 5) are not working correctly. If you use `limit=0` then no entity should be returned but two ones are returned in this case.

Thus, maybe it is not enough with modifying the parameter checking logic in src/lib/rest/rest.cpp and another modifications are needed.

fgalan replied:
A similar situation is happending also in limit_zero_subs.test and limit_zero_regs.test

Anjali-NEC replied:
> This step (and also step 5) are not working correctly. If you use `limit=0` then no entity should be returned but two ones are returned in this case.
> 
> Thus, maybe it is not enough with modifying the parameter checking logic in src/lib/rest/rest.cpp and another modifications are needed.

Done modification in https://github.com/telefonicaid/fiware-orion/pull/3994/commits/cf68c96ee05fa0fd7b53156f8e2b86761a386d0f It works fine in case of limit_zero_subs.test and limit_zero_regs.test but in case of entity it doesn't show fiware-total-count in response. Also I haven't find code for count in getEntities() in src/lib/serviceRoutinesV2/getEntities.cpp. Can you please suggest?

fgalan replied:
> Also I haven't find code for count in getEntities() in src/lib/serviceRoutinesV2/getEntities.cpp. Can you please suggest?

If you search for HTTP_FIWARE_TOTAL_COUNT in the code base you would find the places in which this header is being used so you can see how the calculation is done in each case:

![imagen](https://user-images.githubusercontent.com/1534240/143000149-bd7324e8-8e00-4f67-83b6-85916ca7cac9.png)

In the case of entities I'd say that the first one (postQueryContext.cpp) is the one to look for.

As additional comment, looking to the above results, I'd realized that the `GET /v2/types` is also paginated, so a limit_zero_types.test variants should be also added to this PR.

Anjali-NEC replied:
Added test case for types in https://github.com/telefonicaid/fiware-orion/pull/3994/commits/64cf793d7fae599dcef70fc063ce6302eeb9e5bd

fgalan commented:
Note that the PR is not passing code style check at the present point. It seems the problems is:

```
...
0002 errors of category 'Weird number of spaces at line-start'
=================================================================
...
src/lib/serviceRoutinesV2/getEntities.cpp:308:  Weird number of spaces at line-start.  Are you using a 2-space indent?  [whitespace/indent] [3]
src/lib/serviceRoutinesV2/getEntities.cpp:309:  Weird number of spaces at line-start.  Are you using a 2-space indent?  [whitespace/indent] [3]
...
Total errors found: 17
```

fgalan commented on CHANGES_NEXT_RELEASE:
```diff
@@ -1,3 +1,4 @@
+- Fix: Using limit=0 and options=count to get only the count of elements (#1492)
```
```suggestion
- Fix: allow limit=0 in all paginated operations in the NGSIv2 API (entities, entity types, subscriptions and registrations)  (#1492)
```

Anjali-NEC replied:
Fixed in https://github.com/telefonicaid/fiware-orion/pull/3994/commits/64cf793d7fae599dcef70fc063ce6302eeb9e5bd

Anjali-NEC commented:
> Note that the PR is not passing code style check at the present point. It seems the problems is:
> 
> ```
> ...
> 0002 errors of category 'Weird number of spaces at line-start'
> =================================================================
> ...
> src/lib/serviceRoutinesV2/getEntities.cpp:308:  Weird number of spaces at line-start.  Are you using a 2-space indent?  [whitespace/indent] [3]
> src/lib/serviceRoutinesV2/getEntities.cpp:309:  Weird number of spaces at line-start.  Are you using a 2-space indent?  [whitespace/indent] [3]
> ...
> Total errors found: 17
> ```

Fixed in https://github.com/telefonicaid/fiware-orion/pull/3994/commits/64cf793d7fae599dcef70fc063ce6302eeb9e5bd

fgalan commented:
@Anjali-NEC taking into account the upcoming FIWARE TSC on Orion, could you provide an status update of this PR (or be prepared to attend to the TSC and report at the meeting), please? Thanks!

Anjali-NEC commented:
> @Anjali-NEC taking into account the upcoming FIWARE TSC on Orion, could you provide an status update of this PR (or be prepared to attend to the TSC and report at the meeting), please? Thanks!

In my latest commit https://github.com/telefonicaid/fiware-orion/pull/3994/commits/8a03332e3df26dc5d17d0df34121a6006b0e2955 I have resolved all comments expect only one comment https://github.com/telefonicaid/fiware-orion/pull/3994#discussion_r749387006 which is related to V1 and I think V1 is deprecated so we don't need to fix this. What do you suggest? Although I tried my best to correct it but somehow not able to make it.

fgalan commented:
> > @Anjali-NEC taking into account the upcoming FIWARE TSC on Orion, could you provide an status update of this PR (or be prepared to attend to the TSC and report at the meeting), please? Thanks!
> 
> In my latest commit [8a03332](https://github.com/telefonicaid/fiware-orion/commit/8a03332e3df26dc5d17d0df34121a6006b0e2955) I have resolved all comments expect only one comment [#3994 (comment)](https://github.com/telefonicaid/fiware-orion/pull/3994#discussion_r749387006) which is related to V1 and I think V1 is deprecated so we don't need to fix this. What do you suggest? Although I tried my best to correct it but somehow not able to make it.

I'll have a look when I can.

Anjali-NEC commented:
@fgalan Gentle Reminder !!

fgalan commented on src/lib/serviceRoutines/postQueryContext.cpp:
```diff
@@ -504,14 +505,17 @@ std::string postQueryContext
-                                                      ciP->uriParamOptions,
-                                                      countP,
-                                                      ciP->apiVersion));
+  if (limit != 0)
```
You shouldn't do this check here. Note that you could need the calculation of `countP` done by mongoQueryContext() if the request is something like `GET /v2/entities?limit=0&options=count`. In fact, note step 4 in test/functionalTest/cases/1492_limit_zero/limit_zero_entities.test is not working correctly (you should get `Fiware-Total-Count: 2` instead of `Fiware-Total-Count: 0`).

Instead of that, I'd suggest you do deal with limit=0 situation internally in mongoQueryContext(). Maybe in entitiesQuery() or even in collectionRangedQuery().

Anjali-NEC replied:
Fixed in https://github.com/telefonicaid/fiware-orion/pull/3994/commits/ff2b2251b1f0a0a1703c4e40b51d163f72bcede8

fgalan commented on test/functionalTest/cases/1492_limit_zero/limit_zero_regs.test:
```diff
@@ -0,0 +1,167 @@
+--SHELL--
+
+#
+# 01. GET /v2/registrations with 'limit' == 0 and options=count and see no registrations and fiware-total-count: 0
+# 02. POST /v2/registrations to create registration
+# 03. POST /v2/registrations to create another registration
+# 04. GET /v2/registrations with 'limit' == 0 and options=count and see no registrations and fiware-total-count: 2
+# 05. GET /v2/registrations with 'limit' == 0 and see no registrations
```
In steps 01, 04, 05 you should get `[]` instead of empty payload.

Anjali-NEC replied:
Fixed in https://github.com/telefonicaid/fiware-orion/pull/3994/commits/67ea727c02d09a50bde56fcad40323bd372738c4

fgalan commented on test/functionalTest/cases/1492_limit_zero/limit_zero_subs.test:
```diff
@@ -0,0 +1,179 @@
+--SHELL--
+
+#
+# 01. GET /v2/subscriptions with 'limit' == 0 and options=count and see no subscriptions and fiware-total-count: 0
+# 02. POST /v2/subscriptions to create subscription
+# 03. POST /v2/subscriptions to create another subscription
+# 04. GET /v2/subscriptions with 'limit' == 0 and options=count and see no subscriptions and fiware-total-count: 2
+# 05. GET /v2/subscriptions 'limit' == 0 and see no subscriptions
```
In steps 01, 04, 05 you should get `[]` instead of empty payload.

fgalan commented on test/functionalTest/cases/1492_limit_zero/limit_zero_types.test:
```diff
@@ -0,0 +1,136 @@
+--SHELL--
+
+#
+# 01. GET /v2/types with 'limit' == 0 and options=count and see no types and fiware-total-count: 0
+# 02. POST /v2/entities to create E1 entity with type T1
+# 03. POST /v2/entities to create E2 entity with type T2
+# 04. GET /v2/types with 'limit' == 0 and options=count and see no types and fiware-total-count: 2
+# 05. GET /v2/types with 'limit' == 0 and see no types
```
In steps 01, 04, 05 you should get `[]` instead of empty payload.

Anjali-NEC replied:
Fixed in https://github.com/telefonicaid/fiware-orion/pull/3994/commits/ff2b2251b1f0a0a1703c4e40b51d163f72bcede8

fgalan commented on test/functionalTest/cases/1492_limit_zero/limit_zero_entities.test:
```diff
@@ -0,0 +1,142 @@
+HTTP/1.1 200 OK
+Content-Length: 2
+Content-Type: application/json
+Fiware-Total-Count: 0
```
As I said at https://github.com/telefonicaid/fiware-orion/pull/3994/files#r783052948:

```suggestion
Fiware-Total-Count: 2
```

Anjali-NEC replied:
Fixed in https://github.com/telefonicaid/fiware-orion/pull/3994/commits/ff2b2251b1f0a0a1703c4e40b51d163f72bcede8

fgalan commented on src/lib/serviceRoutinesV2/getAllSubscriptions.cpp:
```diff
@@ -87,7 +87,11 @@ std::string getAllSubscriptions
 
   std::string out;
+
+  if (limit != 0)
+  {
   TIMED_RENDER(out = vectorToJson(subs));
+  }
```
You shouldn't do the limit=0 check here here. This is probably the cause of the problem in .test described in [previous comment](https://github.com/telefonicaid/fiware-orion/pull/3994#discussion_r783055063).

Instead, you should deal with the limit=0 condition internally at mongoListSubscriptions() (or even in collectionRangedQuery()).

Anjali-NEC replied:
Added condition at mongoListSubscriptions()

fgalan commented on src/lib/serviceRoutinesV2/getEntityAllTypes.cpp:
```diff
@@ -84,7 +85,11 @@ std::string getEntityAllTypes
   }
-  TIMED_RENDER(answer = response.toJson(ciP->uriParamOptions[OPT_VALUES]));
+
+  if (limit != 0)
+  {
+    TIMED_RENDER(answer = response.toJson(ciP->uriParamOptions[OPT_VALUES]));
+  }
```
You shouldn't do the limit=0 check here here. This is probably the cause of the problem in .test described in [previous comment](https://github.com/telefonicaid/fiware-orion/pull/3994#discussion_r783055360).

Instead, you should deal with the limit=0 condition internally at mongoEntityTypes().

fgalan commented on src/lib/serviceRoutinesV2/getRegistrations.cpp:
```diff
@@ -87,7 +87,10 @@ std::string getRegistrations
   }
 
-  TIMED_RENDER(out = vectorToJson(registrationV));
+  if (limit != 0)
+  {
+    TIMED_RENDER(out = vectorToJson(registrationV));
+  }
```
You shouldn't do the limit=0 check here here. This is probably the cause of the problem in .test described in [previous comment](https://github.com/telefonicaid/fiware-orion/pull/3994#discussion_r783054794).

Instead, you should deal with the limit=0 condition internally at mongoRegistrationsGet() (or even in collectionRangedQuery()).

Anjali-NEC replied:
Added condition at mongoRegistrationsGet()  in https://github.com/telefonicaid/fiware-orion/pull/3994/commits/67ea727c02d09a50bde56fcad40323bd372738c4

fgalan commented:
> @fgalan Gentle Reminder !!

Please, find a new round of comments provided. Basically, I think you should deal the check for limit 0 at mongoBackend layer and not add serviceRoutine layer. Maybe after fixing that also the V1 ParseError issue gets also solved.

Anjali-NEC commented:
> Please, find a new round of comments provided. Basically, I think you should deal the check for limit 0 at mongoBackend layer and not add serviceRoutine layer. Maybe after fixing that also the V1 ParseError issue gets also solved.

I have done some code changes at mongoBackend layer to check for limit=0.  it works in case of GET subscription and registration but in case of GET entities and types it doesn't show the expected response. 

In case of GET v2/entities with limit=0 it show Fiware-Total-Count: 0 and in case of GET /v2/types the response is empty payload instead of []. I have tried to this but I haven't found suitable solution for this.

Can you please suggest what change need to be done for getting expected response.

fgalan commented on src/lib/serviceRoutinesV2/getAllSubscriptions.cpp:
```diff
@@ -87,6 +87,7 @@ std::string getAllSubscriptions
   }
 
   std::string out;
+
```
Please keep tidy files not affected by the PR.

```suggestion
```

Anjali-NEC replied:
Fixed in https://github.com/telefonicaid/fiware-orion/pull/3994/commits/397ae9e496a9ce542907b10016aa464b69ba71a3

fgalan commented on src/lib/serviceRoutines/postQueryContext.cpp:
```diff
@@ -505,14 +505,14 @@ std::string postQueryContext
-                                                      ciP->uriParamOptions,
-                                                      countP,
-                                                      ciP->apiVersion));
+    TIMED_MONGO(ciP->httpStatusCode = mongoQueryContext(qcrP,
+                                                        qcrsP,
+                                                        ciP->tenant,
+                                                        ciP->servicePathV,
+                                                        ciP->uriParam,
+                                                        ciP->uriParamOptions,
+                                                        countP,
+                                                        ciP->apiVersion));
```
Please avoid breaking existing indentation

```suggestion
  TIMED_MONGO(ciP->httpStatusCode = mongoQueryContext(qcrP,
                                                      qcrsP,
                                                      ciP->tenant,
                                                      ciP->servicePathV,
                                                      ciP->uriParam,
                                                      ciP->uriParamOptions,
                                                      countP,
                                                      ciP->apiVersion));
```

Anjali-NEC replied:
Fixed in https://github.com/telefonicaid/fiware-orion/pull/3994/commits/397ae9e496a9ce542907b10016aa464b69ba71a3

fgalan commented on src/lib/serviceRoutinesV2/getEntityAllTypes.cpp:
```diff
@@ -84,7 +84,6 @@ std::string getEntityAllTypes
                                  totalTypesP,
                                  noAttrDetail));
   }
-  TIMED_RENDER(answer = response.toJson(ciP->uriParamOptions[OPT_VALUES]));
```
Why are you removing this line? From my point of view, it should be kept.

Anjali-NEC replied:
Fixed in https://github.com/telefonicaid/fiware-orion/pull/3994/commits/397ae9e496a9ce542907b10016aa464b69ba71a3

fgalan commented on src/lib/serviceRoutinesV2/getRegistrations.cpp:
```diff
@@ -87,7 +87,7 @@ std::string getRegistrations
   }
 
-  TIMED_RENDER(out = vectorToJson(registrationV));
+    TIMED_RENDER(out = vectorToJson(registrationV));
```
Please avoid breaking existing indentation

```suggestion
  TIMED_RENDER(out = vectorToJson(registrationV));
```

Anjali-NEC replied:
Fixed in https://github.com/telefonicaid/fiware-orion/pull/3994/commits/397ae9e496a9ce542907b10016aa464b69ba71a3

fgalan commented on test/functionalTest/cases/1492_limit_zero/limit_zero_entities.test:
```diff
@@ -0,0 +1,142 @@
+#
+# 01. GET /v2/entities with 'limit' == 0 and options=count and see no entities and fiware-total-count: 0
+# 02. POST /v2/entities to create E1 entity
+# 03. POST /v2/entities to create E1 entity
```
```suggestion
# 03. POST /v2/entities to create E2 entity
```

Anjali-NEC replied:
Fixed in https://github.com/telefonicaid/fiware-orion/pull/3994/commits/397ae9e496a9ce542907b10016aa464b69ba71a3

fgalan commented on src/lib/mongoBackend/mongoQueryTypes.cpp:
```diff
@@ -655,6 +655,7 @@ HttpStatusCode mongoEntityTypes
       *totalTypesP = docs;
       char detailsMsg[256];
       snprintf(detailsMsg, sizeof(detailsMsg), "Count: %u", *totalTypesP);
+
```
Please keep tidy files not affected by the PR.

```suggestion
```

Anjali-NEC replied:
Fixed in https://github.com/telefonicaid/fiware-orion/pull/3994/commits/397ae9e496a9ce542907b10016aa464b69ba71a3

fgalan commented:
I have done a new round of review. The PR looks better than last time. Congrats :)

However, there are new comments that need to be addressed. The main one is this: https://github.com/telefonicaid/fiware-orion/pull/3994#discussion_r818725782, but the other are also important, related with indent, unnecessary whitelines, etc.

fgalan commented:
In adddition, the coflict in CHANGES_NEXT_RELEASE should be solved, so functional tests get another pass.

Anjali-NEC commented:
> I have done a new round of review. The PR looks better than last time. Congrats :)
> 
> However, there are new comments that need to be addressed. The main one is this: [#3994 (comment)](https://github.com/telefonicaid/fiware-orion/pull/3994#discussion_r818725782), but the other are also important, related with indent, unnecessary whitelines, etc.

The two test cases are still failing:
```
----------- Failing tests ------------------
  o  1492_limit_zero/limit_zero_entities.test
  o  1492_limit_zero/limit_zero_types.test
```

fgalan commented:
> > I have done a new round of review. The PR looks better than last time. Congrats :)
> > However, there are new comments that need to be addressed. The main one is this: [#3994 (comment)](https://github.com/telefonicaid/fiware-orion/pull/3994#discussion_r818725782), but the other are also important, related with indent, unnecessary whitelines, etc.
> 
> The two test cases are still failing:
> 
> ```
> ----------- Failing tests ------------------
>   o  1492_limit_zero/limit_zero_entities.test
>   o  1492_limit_zero/limit_zero_types.test
> ```

I have had a look to the limit_zero_entities.test and limit_zero_types.test and they seems to be correct, so it seems there is some fail in your implementation. Could you review it and fix, please?

Anjali-NEC commented:
> I have had a look to the limit_zero_entities.test and limit_zero_types.test and they seems to be correct, so it seems there is some fail in your implementation. Could you review it and fix, please?

Hi @fgalan Sir,

Now all test cases are pass in my latest commit https://github.com/telefonicaid/fiware-orion/pull/3994/commits/ff2b2251b1f0a0a1703c4e40b51d163f72bcede8. Please review it and if it is ok please merge it into master.
Thanks

Anjali-NEC commented:
Hi @fgalan Sir,
Please review my PR and if it is ok please merge it into master.
Thanks

fgalan commented:
Work is not finished, but let's merge in pre-landing branch (issue1492-prelanding) to get easy further analysis and modifications.

fgalan commented:
> Work is not finished, but let's merge in pre-landing branch (issue1492-prelanding) to get easy further analysis and modifications.

Continues in PR https://github.com/telefonicaid/fiware-orion/pull/4104
